### ISSUE TITLE
yolov11 val.py之后的mAP结果和训练完后的mAP结果有差异,都是加载的best.pt,同样的数据，这是为什么呢

### ISSUE THREAD

**Issue by learning123hello** [ultralytics/ultralytics#20295]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

训练结果是：
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95):  95%|█████████▍| 86/91 [00:14<00:00,  8.85it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95):  96%|█████████▌| 87/91 [00:14<00:00,  8.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95):  97%|█████████▋| 88/91 [00:14<00:00,  9.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95):  98%|█████████▊| 89/91 [00:14<00:00,  9.11it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95):  99%|█████████▉| 90/91 [00:14<00:00,  9.22it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 91/91 [00:14<00:00,  6.24it/s]
[169]                    all       2888       2888          1          1      0.995      0.994          1          1      0.995      0.903

mask map是0.903

直接使用val.py脚本进行评测，结果是：
val: Caching images (2.3GB RAM): 100%|██████████| 2888/2888 [00:06<00:00, 450.77it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|█
                   all       2888       2888          1          1      0.995      0.994          1          1      0.995      0.895

mask map是0.895

请问这是什么原因造成的呢？
备注：使用的模型是yolov11n-seg

**Comment by UltralyticsAssistant** [comment]
👋 Hello @learning123hello, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will review your question and assist you soon 😊

**Comment by learning123hello** [comment]
求助，目前没有找到差异所在

**Comment by Y-T-G** [comment]
During training runs, validation with different batch size (2x training batch size).

You should use the same batch sized

**Comment by Y-T-G** [comment]
It shows 91 iterations in your log during first validation, which means batch size of 32.

**Comment by learning123hello** [comment]
Thank you very much for your reply. I tried setting the batch size to 32 when running python val.py (and also experimented with batch sizes of 16 and 64). I noticed that the calculated mAP does vary, but I haven't yet reproduced the mAP obtained during the training process. Additionally, I'd like to ask why the batch size setting affects the final mAP calculation. Looking forward to your response.

非常感谢您的回复，我尝试在python val.py的时候，将batch设置为32.（同时也尝试了batch=16，64），我发现计算出来的map的确会有差异，但是目前还没有复现在训练过程中得到的map，另外我想问下，为什么batch设置的大小会影响最终Map的计算，期待您的回复

**Comment by learning123hello** [comment]
Thank you very much for your reply. I tried setting the batch size to 32 when running python val.py (and also experimented with batch sizes of 16 and 64). I noticed that the calculated mAP does vary, but I haven't yet reproduced the mAP obtained during the training process. Additionally, I'd like to ask why the batch size setting affects the final mAP calculation. Looking forward to your response.

非常感谢您的回复，我尝试在python val.py的时候，将batch设置为32.（同时也尝试了batch=16，64），我发现计算出来的map的确会有差异，但是目前还没有复现在训练过程中得到的map，另外我想问下，为什么batch设置的大小会影响最终Map的计算，期待您的回复 @Y-T-G

**Comment by Y-T-G** [comment]
It's due to minimum rectangle batching with `rect`

https://github.com/ultralytics/ultralytics/issues/2783#issuecomment-2413530537

Did you train with AMP?

**Comment by learning123hello** [comment]
yes，During training, I set this parameter to "true" in the configuration file.
amp: True # (bool) Automatic Mixed Precision (AMP) training, choices=[True, False], True runs AMP check

**Comment by learning123hello** [comment]
The following configurations were set in the configuration file:
seed: 0 # (int) random seed for reproducibility
deterministic: True # (bool) whether to enable deterministic mode
single_cls: False # (bool) train multi-class data as single-class
**rect: False** # (bool) rectangular training if mode='train' or rectangular validation if mode='val'
cos_lr: False # (bool) use cosine learning rate scheduler
close_mosaic: 10 # (int) disable mosaic augmentation for final epochs (0 to disable)
resume: False # (bool) resume training from last checkpoint
**amp: True** # (bool) Automatic Mixed Precision (AMP) training, choices=[True, False], True runs AMP check
fraction: 1.0 # (float) dataset fraction to train on (default is 1.0, all images in train set)
profile: False # (bool) profile ONNX and TensorRT speeds during training for loggers
freeze: None # (int | list, optional) freeze first n layers, or freeze list of layer indices during training
multi_scale: False # (bool) Whether to use multiscale during training
# Segmentation
overlap_mask: True # (bool) merge object masks into a single image mask during training (segment train only)
mask_ratio: 4 # (int) mask downsample ratio (segment train only)
# Classification
dropout: 0.0 # (float) use dropout regularization (classify train only)

# Val/Test settings ----------------------------------------------------------------------------------------------------
val: True # (bool) validate/test during training
split: val # (str) dataset split to use for validation, i.e. 'val', 'test' or 'train'
save_json: False # (bool) save results to JSON file
save_hybrid: False # (bool) save hybrid version of labels (labels + additional predictions)
conf: 0.7 # (float, optional) object confidence threshold for detection (default 0.25 predict, 0.001 val)
iou: 0.5 # (float) intersection over union (IoU) threshold for NMS默认0.7
max_det: 300 # (int) maximum number of detections per image
**half: False** # (bool) use half precision (FP16)
dnn: False # (bool) use OpenCV DNN for ONNX inference
plots: True # (bool) save plots and images during train/val

**Comment by Y-T-G** [comment]
Is the first log from the final validation during training?

**Comment by glenn-jocher** [comment]
Thanks for sharing your configuration details and logs. The discrepancy between training validation (0.903) and standalone validation (0.895) is likely due to several factors:

The key differences are in how validation happens during training versus standalone:

1. During validation within training, the model inherits training settings automatically, while standalone validation uses different defaults that need to be explicitly matched

2. Precision differences: During training you're using AMP (`amp: True`) but not half precision (`half: False`), whereas standalone validation defaults to `half: True`

3. Rectangular batching: You set `rect: False` in your config, but validation mode defaults to `rect: True` when run separately

Try running validation with these exact parameters to match your training conditions:

```
python val.py --data same_data.yaml --weights best.pt --batch 32 --rect False --half False --device 0
```

This should help you get results closer to what you see during training. Small differences may still exist due to how batches are processed when integrated into the training loop versus standalone execution.

**Comment by learning123hello** [comment]
Will validate right away， Thank you for your detailed response.
Additionally, why do we get different results when using different batch sizes during validation?

**Comment by learning123hello** [comment]
I would like to express my sincere gratitude for your response

**Comment by glenn-jocher** [comment]
You're welcome! I'm glad I could help clarify those validation differences. These small variations (<1%) are common in deep learning evaluation pipelines and generally don't impact real-world model performance. Feel free to reach out if you have any other questions about YOLO11 or other Ultralytics models.
